### PR TITLE
feat(ui5-multi-input): Keyboard handling

### PR DESCRIPTION
**Pull request by d3xter666**
- Right, Left, Up, Down
- Right / Left / Up / Down + Shift
- Right / Left / Up / Down + Shift + Ctrl
- Multiple tokens deletion
- Cut/Copy (Ctrl + C/X) tokens functionality 

Related to: #3090

### CODE REVIEW

**Comment by MapTo0** [comment]
Add tests please

File: packages/main/src/Tokenizer.js [hunk 1]
```diff
@@ -259,6 +358,25 @@ class Tokenizer extends UI5Element {
 		}
 	}
 
+	_fillClipboard(sShortcutName, tokens) {
```
Comment (stbodurov) [inline packages/main/src/Tokenizer.js:361]: Hungarian notation is only used is UI5.

**Review by stbodurov — COMMENTED** [review]
When the first list item in a suggestions popover is in focus and **arrow up** is pressed, the initially entered text is not being restored.

SHIFT+CTRL+LEFT/RIGHT do not interact with tokens.

**Comment by d3xter666** [comment]
Tests have been added

File: packages/main/src/Tokenizer.js [hunk 2]
```diff
@@ -259,6 +374,25 @@ class Tokenizer extends UI5Element {
 		}
 	}
 
+	_fillClipboard(shortcutName, tokens) {
+		const tokensTexts = tokens.filter(token => token.selected).map(token => token.text).join("\r\n");
+
+		/* fill clipboard with tokens' texts so parent can handle creation */
+		const cutToClipboard = oEvent => {
```
Comment (stbodurov) [inline packages/main/src/Tokenizer.js:381]: oEvent => event

Comment (stbodurov) [inline packages/main/src/Tokenizer.js:381]:
```suggestion
		const cutToClipboard = event => {
```

File: packages/main/src/Tokenizer.js [hunk 3]
```diff
@@ -259,6 +374,25 @@ class Tokenizer extends UI5Element {
...
+			if (oEvent.clipboardData) {
```
Comment (stbodurov) [inline packages/main/src/Tokenizer.js:382]:
```suggestion
			if (oEvent.clipboardData) {
				oEvent.clipboardData.setData("text/plain", tokensTexts);
			} else {
				oEvent.originalEvent.clipboardData.setData("text/plain", tokensTexts);
			}
```

File: packages/main/src/Tokenizer.js [hunk 4]
```diff
@@ -259,6 +374,25 @@ class Tokenizer extends UI5Element {
...
+				oEvent.clipboardData.setData("text/plain", tokensTexts);
+			} else {
+				oEvent.originalEvent.clipboardData.setData("text/plain", tokensTexts);
+			}
+
+			oEvent.preventDefault();
```
Comment (stbodurov) [inline packages/main/src/Tokenizer.js:388]:
```suggestion
		const cutToClipboard = event => {
			if (event.clipboardData) {
				event.clipboardData.setData("text/plain", tokensTexts);
			} else {
				event.originalEvent.clipboardData.setData("text/plain", tokensTexts);
			}

			event.preventDefault();
```

File: packages/main/src/Tokenizer.js [hunk 5]
```diff
@@ -259,6 +374,25 @@ class Tokenizer extends UI5Element {
...
+		};
```
Comment (stbodurov) [inline packages/main/src/Tokenizer.js:389]:
```suggestion
		const cutToClipboard = event => {
			if (event.clipboardData) {
				event.clipboardData.setData("text/plain", tokensTexts);
			} else {
				event.originalEvent.clipboardData.setData("text/plain", tokensTexts);
			}

			event.preventDefault();
		};
```

**Review by stbodurov — COMMENTED** [review]
Home does not focus VSH when present. If Home is at the start of text, focus should move to first token.

If focus is on the last token, End should move focus to text input.

**Comment by d3xter666** [comment]
> If Home is pressed while caret is at the start of text, focus should move to first token.

Issues have been fixed

File: packages/main/src/Tokenizer.js [hunk 6]
```diff
@@ -237,10 +247,106 @@ class Tokenizer extends UI5Element {
...
+			event.preventDefault();
+
+			const isCut = event.key.toLowerCase() === "x";
+			const selectedTokens = tokens.filter(token => token.selected);
+
+			if (isCut) {
+				selectedTokens.forEach(token => {
+					this.fireEvent("token-delete", { ref: token });
+				});
+				return this._fillClipboard("cut", selectedTokens);
```
[the remark this answers is not in the document]
Comment (d3xter666) [reply packages/main/src/Tokenizer.js]: Done

**Comment by hristop** [comment]
In UI5, If you cut a token / all tokens, the focus goes in the input field, however, here if you do that here, the focus is somewhere else. Pressing Tab brings it into the input field, but it is not initially there.

**Comment by d3xter666** [comment]
> In UI5, If you cut a token / all tokens, the focus goes in the input field, however, here if you do that here, the focus is somewhere else. Pressing Tab brings it into the input field, but it is not initially there.

It's fixed now

**Comment by hristop** [comment]
Approved

**Review by hristop — APPROVED** [review]

**Pull request merged**